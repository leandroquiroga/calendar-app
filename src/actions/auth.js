import { fetchNotToken, fetchWithToken } from "../helpers/fetchAPI"
import { types } from "../types/types";

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

const checkingFinish = () => ({ type: types.authChekingFinish });

const logout = () => ({ type: types.authLogout });

const eventLogout = () => ({ type: types.eventLogout });

export const authStartLogin = (email, password, setMsgError) => {
  return async (dispatch) => {
    const response = await fetchNotToken('auth/login', { email, password }, 'POST');
    const body = await response.json();

    // En caso de que haya un error en el login retorna el mensaje de error.
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());
    dispatch(login({
      uid: body.uid,
      name: body.name
    }));
  }
} 

export const startRegister = (email, password, name, setMsgError) => {
  return async (dispatch) => {
    const response = await fetchNotToken('auth/register', { email, password, name }, 'POST');
    const body = await response.json();

    // En caso de que haya un error en el registro retornara un mensaje de error 
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());
    dispatch(login({
      uid: body.uid,
      name: body.name,
    }));
  };
}


export const startCheking = () => {
  return async (dispatch) => {
    const response = await fetchWithToken('auth/renew');
    const body = await response.json();

    if (!body.ok) {
      // Cambia el checking a true y retorna un mensaje de error desde el backend
      return dispatch(checkingFinish());
    };

    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());

    dispatch(login({
      uid: body.uid,
      name: body.name,
    }));
  };
};


export const startUserLogut = () => {
  return (dispatch) => {
    // Debe eliminar el token
    localStorage.removeItem('token');
    localStorage.removeItem('token-init-date');

    dispatch(eventLogout())
    dispatch(logout());
  }
};