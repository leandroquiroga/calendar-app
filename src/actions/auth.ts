import { Dispatch, SetStateAction } from "react";

import { AppDispatch } from "../store/store";
import { fetchNotToken, fetchWithToken } from "../helpers/fetchAPI"
import { types } from "../types/types";

export interface UserProps {
  email?: string;
  password?: string;
  name: string;
  uid: string;
};

export interface UserLogin {
  email: string;
  password: string;
  setMsgError: Dispatch<SetStateAction<string>>
};

export interface UserRegister {
  email: string;
  password: string;
  name?: string;
  setMsgError: Dispatch<SetStateAction<string>>
};

const login = (user: UserProps) => ({
  type: types.authLogin,
  payload: user
});

const checkingFinish = () => ({ type: types.authChekingFinish });

const logout = () => ({ type: types.authLogout });

const eventLogout = () => ({ type: types.eventLogout });

export const authStartLogin = (
  email: string,
  password: string,
  setMsgError: Dispatch<SetStateAction<string>>
): any => {
  return async (dispatch: AppDispatch) => {
    const response = await fetchNotToken(
      "auth/login",
      { email, password },
      "POST"
    );
    const body = await response.data;

    // En caso de que haya un error en el login retorna el mensaje de error.
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem("token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime().toString());
    dispatch(
      login({
        uid: body.uid,
        name: body.name,
      })
    );
  };
}; 

export const startRegister = (
  email: string,
  password: string,
  name: string,
  setMsgError: Dispatch<SetStateAction<string>>,
): any => {
  return async (dispatch: AppDispatch) => {
    const response = await fetchNotToken(
      "auth/register",
      { email, password, name },
      "POST"
    );
    const body = await response.data;

    // En caso de que haya un error en el registro retornara un mensaje de error
    if (!body.ok) return setMsgError(body.msg);

    localStorage.setItem("token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime().toString());
    dispatch(
      login({
        uid: body.uid,
        name: body.name,
      })
    );
  };
};

// Buscar una mejor definicion para el tipo de retorno de la funcion
export const startCheking = (): any => {
  return async (dispatch: AppDispatch) => {
    const response = await fetchWithToken("auth/renew", "GET");
    const body = await response.data;
    if (!body.ok) {
      // Cambia el checking a true y retorna un mensaje de error desde el backend
      return dispatch(checkingFinish());
    }

    localStorage.setItem("token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime().toString());

    dispatch(
      login({
        uid: body.uid,
        name: body.name,
      })
    );
  };
};

// Buscar una mejor definicion para el tipo de retorno de la funcion
export const startUserLogut = ():any => {
  return (dispatch: AppDispatch) => {
    // Debe eliminar el token
    localStorage.removeItem("token");
    localStorage.removeItem("token-init-date");

    dispatch(eventLogout());
    dispatch(logout());
  };
};