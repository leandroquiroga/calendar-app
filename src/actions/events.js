import { types } from './../types/types';
import { fetchWithToken } from '../helpers/fetchAPI';
import { prepareEvents } from '../helpers/prepareEvents';

const eventAddNew = (event) => ({
  type: types.eventAddNew, 
  payload: event,
});

const eventLoaded = (event) => ({
  type: types.eventLoaded,
  payload: event
});

const updatedEvent = (event) => ({
  type: types.eventUpdateNotes,
  payload: event
});

const deleteEventNotes = () => ({ type: types.eventDeleteNotes });

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {

    const { uid, name } = getState().auth;
    
    try {
      const response = await fetchWithToken('events', event, 'POST');
      const body = await response.json();

      if (body.ok) {
        event.id = body.event.id
        event.user = {
          uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
      
    } catch (error) {
      console.log(error)
    }
  };
};

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
})

export const clearEventActive = () => ({ type: types.clearActiveEvent });

export const eventStartLoading = () => {
  return async(dispatch) => {
    try {
      const response = await fetchWithToken('events');
      const body = await response.json(); 
      const events = prepareEvents(body.event);
      
      dispatch(eventLoaded(events))
    } catch (error) {
      console.log(error)
    }
  };
};

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const body = await response.json();

      if (body.ok)
        return dispatch(updatedEvent(event));

      console.log(body.msg);
    } catch (error) {
      console.log(error)
    }
  }
};

export const eventStartDelete = () => {
  return async (dispatch, getState) => {

    const { id } = getState().calendar.activeEvent;
    
    try {
      const response = await fetchWithToken(`events/${id}`, {}, 'DELETE');
      const body = await response.json();
      
      if (body.ok)
        return dispatch(deleteEventNotes());
      
      console.log(body.msg)
      
    } catch (error) {
      console.log(error)
    }
  }
};