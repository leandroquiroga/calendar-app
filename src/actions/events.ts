import { types } from '../types/types';
import { fetchWithToken } from '../helpers/fetchAPI';
import { AppDispatch, RootState } from '../store/store';
import { convertToEvent } from '../helpers/convertToEvents';
import { prepareEvents } from '../helpers/prepareEvents';

export interface Event {
  title: string;
  notes: string;
  end: string;
  start: string;
  id?: string
  user?: User;
}

export interface PreparedEvent {
  end: string;
  start: string;
  title: string;
  notes: string;
  id?: string;
  user?: User;
};

type User = {
  uid: string;
  name: string;
}
const eventAddNew = (event: Event) => ({
  type: types.eventAddNew, 
  payload: event,
});

const eventLoaded = (event: PreparedEvent[]) => ({
  type: types.eventLoaded,
  payload: event,
});

const updatedEvent = (event: Event) => ({
  type: types.eventUpdateNotes,
  payload: event
});

const deleteEventNotes = () => ({ type: types.eventDeleteNotes });

export const eventStartAddNew = (event: Event): any => {
  return async (dispatch: AppDispatch, getState: () => RootState) => {
    const { uid, name } = getState().auth;
    try {
      const response = await fetchWithToken("events", "POST", event);
      const body = await response.data;

      if (body.ok) {
        event.id = body.event.id;
        event.user = {
          uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventSetActive = (event: Event) => ({
  type: types.eventSetActive,
  payload: event
})

export const clearEventActive = () => ({ type: types.clearActiveEvent });

export const eventStartLoading = (): any => {
  return async (dispatch: AppDispatch) => {
    try {
      const response = await fetchWithToken("events");
      const body = await response.data;
      const preparedEvents = prepareEvents(body.event);
      const events = convertToEvent(preparedEvents);
      dispatch(eventLoaded(events)); 
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartUpdate = (event: Event): any => {
  return async (dispatch: AppDispatch) => {
    try {
      const response = await fetchWithToken(`events/${event.id}`, "PUT", event);
      const body = await response.data;

      if (body.ok) return dispatch(updatedEvent(event));

    } catch (error) {
      console.log(error);
    }
  };
};

// Buscar una mejor definicion para el tipo de retorno de la funcion
export const eventStartDelete = (): any => {
  return async (dispatch: AppDispatch, getState: () => RootState) => {
    const { id } = getState().calendar.activeEvent;

    try {
      const response = await fetchWithToken(`events/${id}`, "DELETE");
      const body = await response.data;

      if (body.ok) return dispatch(deleteEventNotes());

    } catch (error) {
      console.log(error);
    }
  };
};