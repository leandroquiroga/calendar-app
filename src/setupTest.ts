import '@testing-library/jest-dom';
import Enzyme from 'enzyme';
import ReactAdapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer,  } from 'enzyme-to-json';

Enzyme.configure({ adapter: new ReactAdapter() });

expect.addSnapshotSerializer(createSerializer({mode: 'deep'}) as any)