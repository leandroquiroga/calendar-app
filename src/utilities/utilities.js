import confiugureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const initState = {};
export const mockStore = confiugureStore(middlewares);
export let store = mockStore(initState);

export let storeNabvar = mockStore({ auth: { name: 'testing' } });
export let storeModal = mockStore({
  auth: {
    checking: true,
    uid: null
  },
  calendar: {
    event: [],
    activeEvent: {
      title: "Testing",
      notes:"This test should run correctly",
      start:"2022-06-02T12:45:00.000Z",
      end: "2022-06-02T13:00:00.000Z",
      user: {
        _id:"6285455405b1fddeb0b5a999",
        name:"Tester Pro"

      },
      id:"62978ddd74ecd43d8c87ea48",}
  },
  ui: {
    modalOpen: true
  }
})
export let storeAppRouter = mockStore({    
  auth: {
    checking: true,
    uid: null
  },
  calendar: {
    event: [],
    activeEvent: null
  },
  ui: {
    modalOpen: false
  }
});

export let storeAuth = mockStore({
  auth: {
    uid: null
  }
});

