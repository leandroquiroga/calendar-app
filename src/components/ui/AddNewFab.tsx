import React from 'react'
import { useDispatch } from 'react-redux'
import { uiOpenModal } from '../../actions/actions';

export const AddNewFab = (): JSX.Element => {

  const dispatch = useDispatch();
  // Abre el modal 
  const handleOpenModal = () => dispatch(uiOpenModal())
  
  return (
    <button
      className='btn btn-primary shadow rounded-circle button_icon'
      onClick={handleOpenModal}
    >
      <i className="fa-solid fa-plus"></i>
    </button>
  )
};