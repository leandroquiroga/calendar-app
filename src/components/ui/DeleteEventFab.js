import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { deleteEventNotes } from '../../actions/events';


export const DeleteEventFab = () => {

  const dispatch = useDispatch();
  const {activeEvent} = useSelector(state => state.calendar)

  // Elimina el evento
  const handleDeleteNotes = () => dispatch(deleteEventNotes(activeEvent.id))

  return (
    <button
      className='btn btn-danger rounded-circle shadow button_icon__delete'
      onClick={handleDeleteNotes}
    >
        <i className="fa-solid fa-trash-can"></i>
    </button>
  )
}
