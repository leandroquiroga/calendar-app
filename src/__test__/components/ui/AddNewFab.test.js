import { mount } from "enzyme";
import { Provider } from 'react-redux';
import { AddNewFab } from "../../../components/ui/AddNewFab";
import { store } from '../../../utilities/utilities';

import '@testing-library/jest-dom';
import "../../../setupTest";
import { uiOpenModal } from "../../../actions/actions";

jest.mock('../../../actions/actions', () => ({
  uiOpenModal: jest.fn()
})); 

store.dispatch = jest.fn();

const wapper = mount(
  <Provider store={store}>
    <AddNewFab />
  </Provider>
)
describe('Test on <AddNewFab />', () => {
  test('this test should make a to match snapshot with <AddNewFab />', () => {
    expect(wapper).toMatchSnapshot();
  });

  test('this test should simulate the click event of the button', () => {
    wapper.find('button').simulate('click');
    expect(uiOpenModal).toHaveBeenCalled()
  });
})