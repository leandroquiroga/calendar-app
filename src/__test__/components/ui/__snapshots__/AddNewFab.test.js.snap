// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <AddNewFab /> this test should make a to match snapshot with <AddNewFab /> 1`] = `
<button
  className="btn btn-primary shadow rounded-circle button_icon"
  onClick={[Function]}
>
  <i
    className="fa-solid fa-plus"
  />
</button>
`;
